docs/spec/spec.md

```diff
@@ -0,0 +1,1821 @@
+    * 4.6. [Cache config annotation](#46-cache-config-annotation)
+5. [url-parameters](#5-url-parameters)
+    * 5.1. [Path](#51-path)
+    * 5.2. [Query](#52-query)
+    * 5.3. [Matrix](#53-matrix)
+6. [Request and Response](#6-request-and-response)
+7. [Header and Payload](#7-header-and-payload)
+    * 7.1. [Parse header functions](#71-parse-header-functions)
+8. [Interceptor and error handling](#8-interceptor-and-error-handling)
+    * 8.1. [Interceptor](#81-interceptor)
+        * 8.1.1 [Request interceptor](#811-request-interceptor)
+            * 8.1.1.1 [Request context](#8111-requestcontext)
+            * 8.1.1.2 [Next method](#8112-next-method)
+            * 8.1.1.3 [Returning error](#8113-returning-error)
+        * 8.1.2 [Request error interceptor](#812-request-error-interceptor)
+        * 8.1.3 [Engaging interceptor](#813-engaging-interceptors)
+            * 8.1.3.1 [Service level](#8131-service-level)
+            * 8.1.3.2 [Listener level](#8132-listener-level)
+            * 8.1.3.3 [Execution order of interceptors](#8133-execution-order-of-interceptors)
+        * 8.1.4 [Data binding](#814-data-binding)
+        * 8.1.5 [Return to respond](#815-return-to-respond)
+     * 8.2. [Error handling](#82-error-handling)
+        * 8.2.1. [Trace log](#821-trace-log)
+        * 8.2.2. [Access log](#822-access-log)
+9. [Security](#9-security)
+    * 9.1. Authentication and Authorization
+        * 9.1.1. Declarative Approach
+            * 9.1.1.1. Service - File User Store
+            * 9.1.1.2. Service - LDAP User Store
+            * 9.1.1.3. Service - JWT
+            * 9.1.1.4. Service - OAuth2
+            * 9.1.1.5. Client - Basic Auth
+            * 9.1.1.6. Client - Bearer Token Auth
+            * 9.1.1.7. Client - Self Signed JWT Auth
+            * 9.1.1.8. Client - OAuth2
+        * 9.1.2 Imperative Approach
+            * 9.1.2.1. Service - File User Store
+            * 9.1.2.2. Service - LDAP User Store
+            * 9.1.2.3. Service - JWT
+            * 9.1.2.4. Service - OAuth2
+            * 9.1.2.5. Client - Basic Auth
+            * 9.1.2.6. Client - Bearer Token Auth
+            * 9.1.2.7. Client - Self Signed JWT Auth
+            * 9.1.2.8. Client - OAuth2   
+10. [Protocol-upgrade](#10-protocol-upgrade)
+    * 10.1. [HTTP2](#101-http2)
+        * 10.1.1. [Push promise and promise response](#1011-push-promise-and-promise-response)
+    * 10.2. [Websocket](#102-websocket)
+
+## 1. Overview
+This specification elaborates on Basic Auth authentication and authorization for all the Ballerina listeners and
+clients. The HTTP, gRPC, GraphQL, WebSocket, WebSub protocol-based listeners and clients are secured according to this
+specification.
+
+The ballerina language provides the first class support for writing network oriented programming models in a more 
+initiative manner. The ballerina-http standard library consumes the rich language construct and creates the 
+programming model to write services.
```

Comment:
I feel the start can be improved. How about something like this ?

The HTTP standard library is designed to work with HTTP protocol. It includes high level abstractions such as `http:Request`, `http:Response`, `http:Service` and `http:Client` which allow users to consume and produce HTTP services. Further the users can use this library to build other libraries. In fact, the standard libraries such as GraphQL, Websub and WebSubHub use this library internally.  

In addition to functional requirements, this library deals with none functional requirements such as security, observerbility and resiliency. Each requirement is discussed in detail in the coming sections.